### PR TITLE
Direct planner

### PR DESCRIPTION
**Pull request by YLGH**
Summary:
See  https://docs.google.com/document/d/1ASiz0Fu3c8cCHmsbseUdulU4jfTgdwZAxvlywgXiBAU/edit for more context

tl;dr:

Currently the only “feasible” way to specify a sharding plan is by using the OSS planner, and using constraints. I say “feasible” because although we can manually construct a ShardingPlan/ParameterSharding, it’s quite cumbersome to do so. Additionally, users do not have the expressibility to say “I want to place this parameter on rank 0” when using planner.

With this new API we can construct sharding plans in a direct way:

```
ebc = EmbeddingBagCollection(...)
plan = construct_parameter_sharding_plans(
            ebc,
            {
                "table_0": data_parallel_sharding(),
                "table_1": row_wise_sharding(),
                "table_2": column_wise_sharding(),
                "table_3": column_wise_sharding(ranks=[0,1,2]),
                "table_4": table_row_wise_sharding(host_index=2),
            },
            EmbeddingBagCollectionSharder()
        )
ebc = sharder.shard(
   ebc,
   plan
)
```

Note, that this is not intended to replace the greedy planner. Instead it allows users to shard modules without being FORCED to use planner. Users will be able to user planner in the exact same way.

Having alternatives instead of planner has a couple of benefits

1. There are bugs in the planner (we should be able to shard something, but planner may have a false negative rejection)
2. Planner doesn't correctly account for new architectures/models /optimizers
3. Easier to directly use torchrec (new users don't need to understand how to use planner).

-----

So I decided to use a generator function because alternative ways would require creating a new dataclass called like TableWiseConfig(row=...), and I didn't like the look of it.

Having just the atomic generator functions is also nice since it doesn't couple these configs with the construct_parameter_sharding_plans

Differential Revision: D39803183

### CODE REVIEW

File: torchrec/distributed/planner/tests/test_simple_planner.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+
+def _test_sharding(
+    tables: List[EmbeddingBagConfig],
+    initial_state_dict: Dict[str, Any],
+    rank: int,
+    world_size: int,
+    kjt_input_per_rank: List[KeyedJaggedTensor],
+    sharder: ModuleSharder[nn.Module],
+    backend: str,
```
Comment (divchenko) [inline torchrec/distributed/planner/tests/test_simple_planner.py]: nit: group rank, world_size, local size, backend together . Maybe introduce a dataclass so it's easier to pass all the way to MultiProcessContext ?

File: torchrec/distributed/planner/tests/test_simple_planner.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+        feature_keys = []
+        for table in tables:
+            feature_keys.extend(table.feature_names)
+
+        # each rank gets a subbatch
+        sharded_model_pred_kt = sharded_model(kjt_input_per_rank[ctx.rank]).to_dict()
+        _sharded_model_pred = torch.stack(  # noqa
+            [sharded_model_pred_kt[feature] for feature in feature_keys]
+        )
```
Comment (divchenko) [inline torchrec/distributed/planner/tests/test_simple_planner.py]: Do you intend to check something here? e.g. sharded vs unsharded results?

File: torchrec/distributed/planner/planners.py [hunk 3]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+
+    per_parameter_sharding: Dict[str, ParameterSharding] = {}
+
+    for param_name, sharding_plan_generator in per_param_sharding.items():
+        param = None
+        for name, _param in module.named_parameters():
+            if param_name in name:
+                param = _param
+                break
+        assert param is not None
```
Comment (divchenko) [inline torchrec/distributed/planner/planners.py]: better error message

File: torchrec/distributed/planner/planners.py [hunk 4]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+        )
+
+    return _parameter_sharding_generator
+
+
+def construct_parameter_sharding_plans(
+    module: nn.Module,
+    per_param_sharding: Dict[
+        str, Callable[[nn.Parameter, int, int, str], ParameterSharding]
+    ],
```
Comment (divchenko) [inline torchrec/distributed/planner/planners.py]: This Callable is complex and repeated in many places. Define a new type.
ParameterShardingGenerator = Callable[[nn.Parameter, int, int, str], ParameterSharding]

File: torchrec/distributed/planner/planners.py [hunk 5]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+        assert param is not None
+
+        per_parameter_sharding[param_name] = sharding_plan_generator(
+            param, local_size, world_size, device_type
+        )
+
+    return per_parameter_sharding
+
+
+def uniform_sharding(
```
Comment (divchenko) [inline torchrec/distributed/planner/planners.py]: What is the final end-to-end code? We need to list an example somewhere.
```
model = MyModel()
ebc_plan = construct_parameter_sharding_plans(model.ebc, uniform_sharding(model.ebc, row_wise_sharding())
eb_plan = construct_parameter_sharding_plans(model.eb, uniform_sharding(model.eb, row_wise_sharding())
plan = {**eb_plan, **eb_plan}
```

File: torchrec/distributed/planner/planners.py [hunk 6]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+
+    return _parameter_sharding_generator
+
+
+def construct_parameter_sharding_plans(
+    module: nn.Module,
+    per_param_sharding: Dict[
+        str, Callable[[nn.Parameter, int, int, str], ParameterSharding]
+    ],
+    sharder: ModuleSharder[nn.Module],
```
Comment (divchenko) [inline torchrec/distributed/planner/planners.py]: We need to add a higher-level API, where we don't pass sharder. There is a module -> sharder dict. It should operate on top of it.

**Review by divchenko — APPROVED** [review]

File: torchrec/distributed/planner/tests/test_simple_planner.py (hunk 2 above)
Comment (YLGH) [reply torchrec/distributed/planner/tests/test_simple_planner.py]: here I just wanted to check that things run w/ the constructed plans,

there will be another unit test that does correctness validation over composable ebc with inputs from this

File: torchrec/distributed/planner/planners.py (hunk 3 above)
Comment (YLGH) [reply torchrec/distributed/planner/planners.py]: the param_name in name is also a bit annoying because we strip off ".weight", I guess I can get substring bugs

File: torchrec/distributed/planner/planners.py (hunk 5 above)
Comment (YLGH) [reply torchrec/distributed/planner/planners.py]: yeah 
I think there needs to be something that can add FQN path prefix to output of construct_parameter_sharding_plans: Dict[str, ParameterSharding], 

this operates on the Sharder.shard level

File: torchrec/distributed/planner/planners.py [hunk 7]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+
+    return per_parameter_sharding
+
+
+def uniform_sharding(
+    module: nn.Module,
+    parameter_sharding_generator: Callable[
+        [nn.Parameter, int, int, str], ParameterSharding
+    ],
+    sharder: ModuleSharder[nn.Module],
```
Comment (YLGH) [inline torchrec/distributed/planner/planners.py]: default to None and try to find the correct one from dict

File: torchrec/distributed/planner/planners.py (hunk 5 above)
Comment (YLGH) [reply torchrec/distributed/planner/planners.py]:
```
Overall plan would look like
{
   "model.ebc": ebc_plan,
   "model.ec" : ec_plan
}
``` i suppose

File: torchrec/distributed/planner/planners.py [hunk 8]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+                        none_throws(shard.rank),
+                        none_throws(local_size),
+                    ),
+                )
+                for shard in shards
+            ]
+        ),
+        sharding_type=sharding_type,
+        # pyre-ignore
+        compute_kernel=None,
```
Comment (colin2328) [inline torchrec/distributed/planner/planners.py]: setting compute_kernel to None will mean that we cannot use w/ DMP (unless manually changed), IMO add comment here or in user facing API

File: torchrec/distributed/planner/planners.py [hunk 9]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+
+        return _get_parameter_sharding(
+            ShardingType.ROW_WISE.value, shards, local_size, device_type
+        )
+
+    return _parameter_sharding_generator
+
+
+def column_wise_sharding(
+    shard_dim: int = 128,
```
Comment (colin2328) [inline torchrec/distributed/planner/planners.py]: add docstrings, including what shard_dim represents

File: torchrec/distributed/planner/planners.py [hunk 10]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+) -> Dict[str, ParameterSharding]:
+    """
+    Helper function to create sharding plans (Dict[str, ParameterSharding]) for an module
+    Args:
+        module (nn.Module): module to create plan for.
+        per_param_sharding: Dict[str, Callable[[nn.Parameter, int, int, str], ParameterSharding]]: A mapping of parameter names to a generator function
+        that takes in [parameter, local_size, world_size, device_type] and returns a ParameterSharding. We recommend using one of the predefined generator functions
+        e.g. table_wise_sharding, row_wise_sharding, etc,
+        sharder: ModuleSharder[nn.Module]: Sharder that we are creating a plan for,
+        local_size: Optional[int] = None: process group local size
```
Comment (colin2328) [inline torchrec/distributed/planner/planners.py]: add generally user should not pass these in

File: torchrec/distributed/planner/planners.py [hunk 11]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+        world_size: Optional[int] = None: process_group world_size
+        device_type: str : Torch device type,
```
Comment (colin2328) [inline torchrec/distributed/planner/planners.py]: IMO we should use device, or at least torch.device.type rather than str

File: torchrec/distributed/planner/planners.py [hunk 12]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+
+    Example::
+        ebc = EmbeddingBagCollection(...)
+        plan = construct_parameter_sharding_plan(
+            ebc,
+            {
+                "table_0": data_parallel_sharding(),
+                "table_1": row_wise_sharding(),
+                "table_2": column_wise_sharding(),
+                "table_3": column_wise_sharding(ranks=[0,1,2]),
```
Comment (colin2328) [inline torchrec/distributed/planner/planners.py]: consider allowing this, or if not at least update to reflect (need to specify shard_dim)

File: torchrec/distributed/planner/planners.py [hunk 13]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+                "table_4": table_row_wise_sharding(host_index=2),
```
Comment (colin2328) [inline torchrec/distributed/planner/planners.py]: we need to write it or comment that it's not supported

File: torchrec/distributed/planner/planners.py [hunk 14]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+
+    local_size = local_size or get_local_size()
+    world_size = world_size or dist.get_world_size()
+
+    per_parameter_sharding: Dict[str, ParameterSharding] = {}
+
+    for param_name, sharding_plan_generator in per_param_sharding.items():
+        param = None
+        for name, _param in module.named_parameters():
+            if param_name in name:
```
Comment (colin2328) [inline torchrec/distributed/planner/planners.py]: we need to strip weight and actually match rather than relying on substring match

Comment (colin2328) [reply torchrec/distributed/planner/planners.py]: since you already do assertion above just iterate through and do .get()

File: torchrec/distributed/planner/planners.py [hunk 15]
```diff
@@ -302,3 +300,267 @@ def plan(
...
+    ],
+    sharder: ModuleSharder[nn.Module],
+) -> Dict[str, Callable[[nn.Parameter, int, int, str], ParameterSharding]]:
+    """
+    Convenience function to generate a uniform sharding scheme for construct_parameter_sharding_plans.
+
+    Example::
+    ebc = EmbeddingBagCollection(...)
+    sharder = EmbeddingBagCollectionSharder()
+    plan = construct_parameter_sharding_plan(
```
Comment (colin2328) [inline torchrec/distributed/planner/planners.py]: IMO we should avoid uniform and instead call it something that indicates were sharding all tables like this "shard_all_tables"

**Comment by colin2328** [comment]
nice
